File: core/src/main/java/com/lomekwi/cave/resource/media/MediaFactory.java
package com.lomekwi.cave.resource.media;

import com.lomekwi.cave.util.MimeType;
import org.bytedeco.javacv.FFmpegFrameGrabber;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

public class MediaFactory {
    private final Map<String, Function<String, MedRes>> map = new HashMap<>();

    public MediaFactory() {
        map.put("video/*", VdoRes::new);
        map.put("audio/*", AudRes::new);
        map.put("image/*", ImgRes::new);
    }

    public MedRes create(String mimeType, String path){
        Function<String, MedRes> constructor = findConstructor(mimeType);
        if (constructor == null) {
            throw new IllegalArgumentException("Unsupported mime type: " + mimeType);
        }
        return constructor.apply(path);
    }

    /**
     * 为一个文件创建所有可用的媒体资源。
     * 视频文件如果包含音频流，会额外创建 AudRes。
     */
    public List<MedRes> createAll(String mimeType, String path) {
        List<MedRes> results = new ArrayList<>();

        String typeWildcard = MimeType.getTypeWildcard(mimeType);
        Function<String, MedRes> constructor = findConstructor(mimeType);

        if (constructor == null) {
            throw new IllegalArgumentException("Unsupported mime type: " + mimeType);
        }

        // 主资源
        results.add(constructor.apply(path));

        // 视频文件如果包含音频流，额外创建 AudRes
        if (typeWildcard.equals("video/*") && hasAudioStream(path)) {
            results.add(new AudRes(path));
        }

        return results;
    }

    /**
     * 轻量探测文件是否包含音频流
     */
    private static boolean hasAudioStream(String path) {
        try (FFmpegFrameGrabber g = new FFmpegFrameGrabber(path)) {
            g.start();
            return g.getAudioChannels() > 0;
        } catch (Exception e) {
            return false;
        }
    }

    private Function<String, MedRes> findConstructor(String mimeType) {
        Function<String, MedRes> constructor = map.get(mimeType);
        if (constructor != null) {
            return constructor;
        }

        String typeWildcard = MimeType.getTypeWildcard(mimeType);
        return map.get(typeWildcard);
    }
    public boolean isSupported(String mimeType){
        if (mimeType == null) return false;
        return findConstructor(mimeType) != null;
    }
    public void register(String mimeType, Function<String, MedRes> constructor){
        map.put(mimeType, constructor);
    }
    public void unregister(String mimeType){
        map.remove(mimeType);
    }
}

File: core/src/main/scala/com/lomekwi/cave/resource/media/MediaFactory.scala
package com.lomekwi.cave.resource.media

import com.lomekwi.cave.util.MimeType
import org.bytedeco.javacv.FFmpegFrameGrabber

import scala.collection.mutable

class MediaFactory {
  import MediaFactory.*

  private val constructors: mutable.Map[String, String => MedRes] = mutable.Map(
    "video/*" -> (path => new VdoRes(path)),
    "audio/*" -> (path => new AudRes(path)),
    "image/*" -> (path => new ImgRes(path))
  )

  def create(mimeType: String, path: String): MedRes = {
    val constructor = findConstructor(mimeType).getOrElse {
      throw new IllegalArgumentException("Unsupported mime type: " + mimeType)
    }
    constructor(path)
  }

  /**
   * 为一个文件创建所有可用的媒体资源。
   * 视频文件如果包含音频流，会额外创建 AudRes。
   */
  def createAll(mimeType: String, path: String): List[MedRes] = {
    val typeWildcard = MimeType.getTypeWildcard(mimeType)
    val constructor = findConstructor(mimeType).getOrElse {
      throw new IllegalArgumentException("Unsupported mime type: " + mimeType)
    }

    if (typeWildcard.equals("video/*") && hasAudioStream(path)) {
      List(constructor(path), new AudRes(path))
    } else {
      List(constructor(path))
    }
  }

  private def findConstructor(mimeType: String): Option[String => MedRes] = {
    constructors.get(mimeType).orElse(constructors.get(MimeType.getTypeWildcard(mimeType)))
  }

  def isSupported(mimeType: String): Boolean = {
    mimeType != null && findConstructor(mimeType).isDefined
  }

  def register(mimeType: String, constructor: String => MedRes): Unit = {
    constructors.put(mimeType, constructor)
  }

  def unregister(mimeType: String): Unit = {
    constructors.remove(mimeType)
  }
}

object MediaFactory {
  private def hasAudioStream(path: String): Boolean = {
    val g = new FFmpegFrameGrabber(path)
    try {
      g.start()
      g.getAudioChannels > 0
    } catch {
      case _: Exception => false
    } finally {
      g.close()
    }
  }
}
